cirq-core/cirq/transformers/routing/route_circuit_cqc.py

```diff
@@ -0,0 +1,327 @@
+# limitations under the License.
+
+"""Heuristic qubit routing algorithm based on arxiv:1902.08091."""
+
+from typing import List, Optional, Set, Tuple, TYPE_CHECKING
+from itertools import combinations
+import networkx as nx
+
+from cirq import circuits, ops, protocols
+from cirq.transformers import transformer_api, transformer_primitives
+from cirq.transformers.routing import mapping_manager, line_initial_mapper
+
+if TYPE_CHECKING:
+    import cirq
+
+
+@transformer_api.transformer
+class RouteCQC:
+    """Transformer class that implements a circuit routing algorithm.
+
+    The algorithm proceeds as follows:
+        (1) Compute the timesteps of the circuit: considering operations in the given circuit from
+            beginning to end, the next timestep is a maximal set of 2-qubit operations that act on
+            disjoint qubits. It is 'maximal' because any 2-qubit gate's qubits in the next timestep
+            must intersect with the qubits that are acted on in the current timestep.
+
+        (2) Places the logical qubits in the input circuit onto some input device by using an
+            initial mapper (LineInitialMapper by default).
+
+        (3) Insert necessary swaps to ensure all 2-qubit gates are executable on the device by
+            traversing the timesteps from left to right while:
+                (i) Removing any single qubit gate and executable 2-qubit gate in the current
+                    timestep and add it to the output routed circuit.
+               (ii) If there aren't any gates left in the current timestep, move on to the next.
+              (iii) If there are gates remaining in the current timesteps, consider a set of
+                    candidate swaps on them and rank them based on a heuristic cost function. Pick
+                    the swap that minimises the cost and use it to update our logical to physical
+                    mapping. Repeat from (i).
+    """
+
+    def __init__(self, device_graph: nx.Graph):
+        """Initializes the circuit routing transformer."""
+
+        if nx.is_directed(device_graph):
+            raise ValueError("Device graph must be undirected.")
+        self.device_graph = device_graph
+
+    def __call__(
+        self,
+        circuit: 'cirq.AbstractCircuit',
+        *,
+        max_search_radius: int = 8,
+        preserve_moment_structure: bool = False,
+        tag_inserted_swaps: bool = False,
+        initial_mapper: Optional['cirq.AbstractInitialMapper'] = None,
+        context: Optional['cirq.TransformerContext'] = None,
+    ) -> 'cirq.AbstractCircuit':
+        """Transforms the given circuit to make it executable on the device.
+
+        Since routing doesn't necessarily modify any specific operation and only adds swaps
+        before /after operations to ensure the circuit can be executed, tagging operations with
+        tags from context.tags_to_ignore will have no impact on the routing procedure.
+
```

Comment:
The documentation should explain:
1) What does preserve moment structure mean? 
2) How are we dealing with `context.deep` + circuit operations?